bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/api/MycroftConnection.java

```diff
@@ -0,0 +1,268 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.mycroft.internal.api;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
+import org.eclipse.jetty.websocket.api.annotations.WebSocket;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
+import org.openhab.binding.mycroft.internal.api.dto.BaseMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * Establishes and keeps a websocket connection to the Mycroft bus
+ *
+ * @author Gwendal ROULLEAU - Initial contribution. Inspired by the deconz binding.
+ */
+@WebSocket
+@NonNullByDefault
+public class MycroftConnection {
+    private static final AtomicInteger INSTANCE_COUNTER = new AtomicInteger();
+    private final Logger logger = LoggerFactory.getLogger(MycroftConnection.class);
+
+    private final WebSocketClient client;
+    private final String socketName;
+    private final Gson gson;
+
+    private final MycroftConnectionListener connectionListener;
+    private final Map<MessageType, Set<MycroftMessageListener<? extends BaseMessage>>> listeners = new ConcurrentHashMap<>();
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private @Nullable Session session;
+
+    private static final int TIMEOUT_MILLISECONDS = 3000;
+
+    public MycroftConnection(MycroftConnectionListener listener, WebSocketClient client) {
+        this.connectionListener = listener;
+        this.client = client;
+        this.client.setConnectTimeout(TIMEOUT_MILLISECONDS);
+        this.client.setMaxIdleTimeout(0);
+        this.socketName = "Websocket$" + System.currentTimeMillis() + "-" + INSTANCE_COUNTER.incrementAndGet();
```

Comment:
I would at least encode the binding name "mycroft" in the socket name since the ´WebSocketClient` is a global one in openHAB for all services.

Reply:
Done